domain/secretbackend/service/service.go

```diff
@@ -0,0 +1,580 @@
+)
+
+// SecretProviderRegistry is a function that returns a secret backend provider for the given backend type.
+type SecretProviderRegistry func(backendType string) (provider.SecretBackendProvider, error)
+
+type Service struct {
+	st             State
+	logger         Logger
+	clock          clock.Clock
+	controllerUUID string
+	registry       SecretProviderRegistry
+}
+
+// NewService creates a new Service for interacting with the secret backend state.
+func NewService(
+	st State, logger Logger,
+	controllerUUID string,
+	clk clock.Clock,
+	registry SecretProviderRegistry,
+) *Service {
+	return &Service{
+		st:             st,
+		logger:         logger,
+		controllerUUID: controllerUUID,
+		clock:          clk,
+		registry:       registry,
+	}
+}
+
+func getK8sBackendConfig(cloud cloud.Cloud, cred cloud.Credential) (*provider.BackendConfig, error) {
+	spec, err := cloudspec.MakeCloudSpec(cloud, "", &cred)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	k8sConfig, err := kubernetes.BuiltInConfig(spec)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return k8sConfig, nil
+}
+
+// GetSecretBackendConfigForAdmin returns the secret backend configuration for the given backend ID for an admin user.
+func (s *Service) GetSecretBackendConfigForAdmin(
+	ctx context.Context, modelUUID coremodel.UUID, cloud cloud.Cloud, cred cloud.Credential,
+) (*provider.ModelBackendConfigInfo, error) {
+	var info provider.ModelBackendConfigInfo
+	m, err := s.st.GetModel(ctx, modelUUID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	info.Configs = make(map[string]provider.ModelBackendConfig)
+	// We need to include builtin backends for secret draining and accessing those secrets while drain is in progress.
+	// TODO(secrets) - only use those in use by model
+	// For now, we'll return all backends on the controller.
+	jujuBackendID := s.controllerUUID
+	info.Configs[jujuBackendID] = provider.ModelBackendConfig{
+		ControllerUUID: s.controllerUUID,
+		ModelUUID:      m.ID.String(),
+		ModelName:      m.Name,
+		BackendConfig:  juju.BuiltInConfig(),
+	}
+	currentBackend, err := s.st.GetSecretBackend(ctx, secretbackend.BackendIdentifier{ID: m.SecretBackendID})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if currentBackend.Name == provider.Auto || currentBackend.Name == provider.Internal {
+		// TODO: The logic for determining the active backend ought to be relocated to an
+		// independent function. This function should then be invoked either when the model
+		// is initially created or when the model's secret backend configuration is reset
+		//  due to the current backend being forcefully removed. Under these circumstances,
+		// the model secret backend will never be set to `auto`; so here we can
+		// simply return the value stored in the database.
+		info.ActiveID = jujuBackendID
+	}
+
+	if m.Type == coremodel.CAAS {
+		k8sConfig, err := getK8sBackendConfig(cloud, cred)
```

Comment:
This isn't always true. You can have a k8s model which has been configured to user vault for example. We are trying to design not to use "model.type == caas" any more. We should just be using the same logic to look at the backed associated with a model via the join table and asking the configured provider factory to create it.

Reply:
This is just simple copied from the facade, and all these extra logic for handling the k8s and internal backends will be gone once the above TODO is resolved.